src/ppom-storage.ts

```diff
@@ -0,0 +1,208 @@
+   * 3. Remove the files that are not up to date in the cdn.
+   * 4. Remove the files that are not in the local storage from the metadata.
+   * 5. Delete the files that are not in the metadata from the local storage.
+   *
+   * @param versionInfo - Version information of metadata files.
+   */
+  async syncMetadata(versionInfo: FileInfo[]): Promise<PPOMFileMetadata> {
+    const metadata = this.#readMetadata();
+    const syncedMetadata = [];
+
+    for (const fileMetadata of metadata) {
+      // check if the file is readable (e.g. corrupted or deleted)
+      try {
+        await this.readFile(fileMetadata.name, fileMetadata.chainId);
+      } catch (exp: any) {
+        continue;
+      }
+
+      // check if the file exits and up to date in the storage
+      if (
+        !versionInfo.find(
+          (file) =>
+            file.name === fileMetadata.name &&
+            file.chainId === fileMetadata.chainId &&
+            file.version === fileMetadata.version &&
+            file.checksum === fileMetadata.checksum,
+        )
+      ) {
+        continue;
+      }
+
+      syncedMetadata.push(fileMetadata);
+    }
+
+    const filesInDB = await this.#storageBackend.dir();
+    for (const { name, chainId } of filesInDB) {
+      if (
+        !syncedMetadata.find(
+          (file) => file.name === name && file.chainId === chainId,
+        )
+      ) {
+        await this.#storageBackend.delete({ name, chainId });
+      }
+    }
+
+    this.#writeMetadata(syncedMetadata);
+    return syncedMetadata;
+  }
+
+  /**
+   * Read the file from the local storage.
+   * 1. Check if the file exists in the local storage.
+   * 2. Check if the file exists in the metadata.
+   * 3. Check if the checksum is valid.
+   *
+   * @param name - Name assigned to storage.
+   * @param chainId - ChainId for which file is queried.
+   */
+  async readFile(name: string, chainId: string): Promise<ArrayBuffer> {
+    const metadata = this.#readMetadata();
+    const fileMetadata = metadata.find(
+      (file) => file.name === name && file.chainId === chainId,
+    );
+    if (!fileMetadata) {
+      throw new Error(
+        `File metadata not found for File (${name}, ${chainId}) not found`,
+      );
+    }
+
+    const data = await this.#storageBackend.read(
+      { name, chainId },
+      fileMetadata.checksum,
+    );
+    if (!data) {
+      throw new Error(`Storage File (${name}, ${chainId}) not found`);
+    }
+
+    return data;
+  }
+
+  /**
+   * Write the file to the local storage.
+   * 1. Check if the checksum is valid.
+   * 2. Write the file to the local storage.
+   * 3. Update the metadata.
+   *
+   * @param options - Object passed to write to storage.
+   * @param options.data - File data to be written.
+   * @param options.name - Name to be assigned to the storage.
+   * @param options.chainId - Current ChainId.
+   * @param options.version - Version of file.
+   * @param options.checksum - Checksum of file.
+   */
```

Comment:
> 1. Check if the checksum is valid.

I'm having trouble finding where this check happens. Where is the checksum checked?

Reply:
Good point, I updated the comment.

Reply:
The check was there previously, later I moved it to indexDB implementation.